### PR TITLE
JS Tracker, New method `resetUserId` to de-assign a user id after a logout 

### LINKED ISSUE #7556: UserID "Signing out use-case" - actions still attributed to the same Visitor (2.12.0)
The Signing-out use-case in the Docs states: "In the case where a visitor visits your website and is logged-in (User ID is set) then a visit will be created for this User ID. If then visitor logs out then the visitor has no User ID set any more and a new visit will be created on the page view right after logging out. (if User ID was not used then the requests would have been tracked into one visit and one unique visitor)."

This seems to be not the case. When a user logs out (User-Id will not be set at all (not even empty string) anymore after logging out), all actions are still attributed to the visitor that had the UserId set, if the action is within "visit_standard_length" (or window_look_back_for_visitor). After waiting for timeout of "visit_standard_length", a new visitor without UserId is created, which would be expected even within "visit_standard_length".

### PR DESCRIPTION
**Pull request by znerol**
Replicates the user-signout mechanism from the PHP tracker in order to fix #7556

### CODE REVIEW

**Comment by znerol** [comment]
Tests pass except for the screenshot thing. Is that a signal or noise?

**Comment by 1u** [comment]
nice!

**Comment by mattab** [comment]
> Tests pass except for the screenshot thing. Is that a signal or noise?

it's noise

**Comment by mattab** [comment]
Thanks for creating this Pull request @znerol :+1: 

Feedback:

* This PR is solving a very important use case for which we currently don't have a good solution: logging out users in SPA (currently the User Id stays in subsequent pageviews in SPA, even after logout..)
* rename `setNewVisitorId`: since User ID logout use case in Single Page Apps is the main use case  for this method, maybe we call it `resetUserId` (or `unsetUserId`, or even `logoutUserId`) if anyone has some thoughts?
* in the `setUserId` method, let's not call the new method when user id is false. It's best to force users to call `resetUserId` separately when needed. 

@znerol feel free to make the changes, otherwise we will likely work on this next month.

**Comment by znerol** [comment]
Update PR, I've opted for `resetUserId`. It not only clears the user id but also generates a new visitor id. `unsetUserId` does not really capture that bit. Also `logoutUserId` could be misleading, we are not talking about privilege levels here.

**Comment by znerol** [comment]
Should I squash the commits? (Except for the minify piwik.js one.)

File: tests/javascript/index.php [hunk 1]
```diff
@@ -3170,12 +3171,12 @@ function getVisitorIdFromCookie(tracker) {
 
 
         // Verify that when resetting the User ID, it also changes the Visitor ID
-        tracker.setUserId(false);
-        ok(getVisitorIdFromCookie(tracker).length == 16, "after setting empty user id, visitor ID from cookie should still be 16 chars, got: " + getVisitorIdFromCookie(tracker));
-        equal(getVisitorIdFromCookie(tracker), visitorId, "after setting empty user id, visitor ID from cookie should be the same as previously ("+ visitorId +")");
+        tracker.resetUserId();
```
Comment (mattab) [inline tests/javascript/index.php]: Here could you also test that getUserId is empty string ?

**Comment by mattab** [comment]
Thanks for updating the PR! fyi: It's not needed to squash commits as when we merge it will squash them. 

Code review
*  left a comment. 
* feel free to add the new feature to CHANGELOG.md but we'll do it otherwise
* Looks good to me otherwise! Maybe @tsteur you can take a quick look?

after merge will need to update the documentation in: https://developer.matomo.org/api-reference/tracking-javascript
  * and https://developer.matomo.org/guides/tracking-javascript-guide#user-id
  * and update our single page app guide here to mention ability to resetUserId after logout https://matomo.org/blog/2017/02/how-to-track-single-page-websites-using-piwik-analytics/

**Comment by mattab** [comment]
-> Check out the documentation for this new feature here: https://developer.matomo.org/guides/tracking-javascript-guide#when-user-logs-out-reset-user-id

it is available in Matomo 3.4.0 due to release today :tada:

**Pull request merged**